### DISCUSSION TITLE
Why Compiler Error CS0416 exist?

### DISCUSSION THREAD

**Discussion by soroshsabz** [dotnet/csharplang#7252] [Q&A]
ITNOA

as you see in https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0416

'type parameter': an attribute argument cannot use type parameters

A type parameter was used as an attribute argument, which is not allowed. Use a non-generic type.

The following sample generates CS0416:

```
// CS0416.cs  
public class MyAttribute : System.Attribute  
{  
   public MyAttribute(System.Type t)  
   {  
   }  
}  
  
class G<T>  
{  
  
   [MyAttribute(typeof(G<T>))]  // CS0416  
   public void F()  
   {  
   }  
  
}
```

My question is, why we cannot pass typeof generic type to attribute?

thanks

**Comment by TahirAhmadov** [comment]
This is probably because the type information has to be encoded somehow to be stored in the assembly, and there are technical reasons why encoding a generic argument type is difficult or impossible. There was a discussion about this a few months ago (no way I can find it now); there is a feasible way to make this work, but it requires some heavy lifting WRT both runtime and language changes, and I don't think it's high priority enough to work on right now.

**Reply by soroshsabz** [reply]
Any update in new C#?

**Reply by jnm2** [reply]
ECMA-335 §II.23.3 on attribute argument encoding:

> If the parameter kind is _System.Type_, (also, the middle line in above diagram) its value is stored as a _SerString_ (as defined in the previous paragraph), representing its canonical name. The canonical name is its full type name, followed optionally by the assembly where it is defined, its version, culture and public -key-token. If the assembly name is omitted, the CLI looks first in the current assembly, and then in the system library (mscorlib); in these two special cases, it is permitted to omit the assembly-name, version, culture and public-key-token.

The "full type name" of `List<int>` in this example would be ``System.Collections.Generic.List`1[[System.Int32, System.Runtime, Version=10.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]``.

The "full type name" of `List<T>` would have to be ``System.Collections.Generic.List`1[!0]``, where `!0` would force the name to be interpreted with a generic context of the type definition of `class C<T>`, referring to `C`'s `T` type parameter. ECMA-335 (§II.9) specifies how generic type parameters are encoded in generic type names:

> Within a generic type definition, its generic parameters are referred to by their index. Generic parameter zero is referred to as `!0`, generic parameter one as `!1`, and so on. Similarly, within the body of a generic method definition, its generic parameters are referred to by their index; generic parameter zero is referred to as `!!0`, generic parameter one as `!!1`, and so on.

So it seems clear what the metadata encoding would be. I haven't found where ECMA-335 says that a System.Type attribute value's SerString _must not_ contain generic parameters.

**Comment by hez2010** [comment]
Attributes are stored as constants in assembly, so it is impossible to use a type parameter that cannot be resolved at compile time.

**Reply by luizfbicalho** [reply]
> Attributes are stored as constants in assembly, so it is impossible to use a type parameter that cannot be resolved at compile time.

But I don't understand is that when you use the generic class, you can resolve it, and it's compile time for that resolution. now that we have the generic attributes in place I think that this could be done

**Reply by bernd5** [reply]
https://github.com/dotnet/csharplang/blob/98d6837c32e8d0ab25a29001267be5be206a0f19/proposals/csharp-11.0/generic-attributes.md

**Reply by colejohnson66** [reply]
> [..] it's compile time for that resolution.

No, generics are resolved at *runtime*. *Type propagation* is done at compile time, but that's not the same thing.

**Reply by luizfbicalho** [reply]
I understand the fact that generics are resolved in runtime, but I don't understand why is that

**Reply by AartBluestoke** [reply]
> I understand the fact that generics are resolved in runtime, but I don't understand why is that

because in general the type can be anything, and potentially not even known at runtime. The compilation of a generic class produces a generic compiled object that is shared between instantiations - List<templatedThing>  - and the compile time doesn't have the concept of a separate List<int> and List<string> - the jit subs in 'int' and 'string' into the single template as needed - more imporantly, using List in your code, doesn't add 'list' code to your binary beyond linking to the library

**Reply by luizfbicalho** [reply]
that part is OK, and the part that `List<>` is the same used in `List<int>` and `List<string>` is ok

but `List<int> != List<string>` and I don't get why `typeof(List<int>).GetCustomAttributes()` isn't diffenrent from `typeof(List<string>).GetCustomAttributes()`


Or shoudn't

**Reply by AartBluestoke** [reply]
becaues List<int> only exists at runtime, at compile time they are all List<~1>, 
typeof(List<int>) compiles to typeof(<the compiled List thingy with 1 template parameter>) - which is the same for string,

Someone with more runtime knowledge than me can comment on why the runtime isn't able to communicate the difference here, but i suspect the answer is 'because that is how it was originally coded and changing this would break so much legacy code'

**Reply by jnm2** [reply]
What kind of legacy code break are you thinking of?
